### ISSUE TITLE
add param into string function

### ISSUE THREAD

**Issue by goatshriek** [goatshriek/stumpless#451]
A function exists to get the string representation of a param, `stumpless_param_to_string`, which returns a new dynamically-allocated string representing the param. However, other functions would benefit from getting this representation directly into a buffer, avoiding the dynamic memory allocation (and subsequent free). This change adds a function that supports this.

## General Approach
There are a few details left out of the following approach, for you to fill in as you encounter them. If you find you need help, please ask here or on the [project gitter](https://gitter.im/stumpless/community) and someone can help you get past the stumbling block.

First, read [this](https://github.com/goatshriek/stumpless/blob/latest/docs/development.md#adding-new-functions) section of the development documentation on adding new functions. Once you understand how this process works, start defining the new function.

Review the existing implementation of `stumpless_param_to_string` in `src/param.c`. Once you understand how the current implementation works, start adding the new functionality to the same file and the associated header, in a function named `stumpless_param_into_string`, with the following signature:

```c
size_t
stumpless_param_into_string( const struct stumpless_param *param, char *str, size_t max_size );
```

The function will write the string representation into the buffer given by `str`, up to `max_size` including a NULL terminator character. It will return the size of the string that would be written into the buffer. Note the difference between the length and size of a string: strings with multibyte UTF-8 characters have a size greater than 1. If the required size is greater than the max_size (again, both including the terminating NULL character) then nothing is copied into the buffer, an error is raised, and the required size is returned.

You'll need to define a new error for this: `STUMPLESS_ARGUMENT_TOO_SMALL`. These are defined in `include/stumpless/error.h`, and supporting functions to raise errors in `include/private/error.h` and `src/error.c`. You'll need to add error message and code type strings for this as well. The [localization docs](https://github.com/goatshriek/stumpless/blob/latest/docs/localization.md) describe how to add new strings that support translation.

Add tests for your new functionality in the existing test module `test/function/param.cpp`, based on the tests already there. Make sure that you have full coverage on your new function, and that you test for multibyte characters.

**Comment by faresbouzayen** [comment]
Hi @goatshriek! I’m excited to take on the task of adding the stumpless_param_into_string function! 🎉

Approach:
Understanding Existing Functionality:

I’ll start by reviewing the stumpless_param_to_string function in src/param.c to understand its current implementation and how to adapt it for the new function.
Implementing the New Function:

I will define stumpless_param_into_string with the signature:
c
Copy code
size_t stumpless_param_into_string(const struct stumpless_param *param, char *str, size_t max_size);
This function will copy the string representation of the parameter into the provided buffer, handling multibyte UTF-8 characters appropriately.
Error Handling:

I’ll introduce the new error type STUMPLESS_ARGUMENT_TOO_SMALL in include/stumpless/error.h and ensure it raises an appropriate error when the buffer is too small.
Localization:

I’ll add the necessary error message and code strings for localization as per the documentation.
Testing:

I’ll create tests for this new function in test/function/param.cpp, ensuring comprehensive coverage, especially for multibyte character cases.
Next Steps:
I’ll read through the development documentation on adding new functions and start implementing the changes. If I encounter any issues, I’ll reach out here or on the project Gitter.
Looking forward to making this enhancement! Let me know if there are any specific guidelines or additional considerations you’d like me to keep in mind. Cheers! 🚀

**Comment by goatshriek** [comment]
Ignore all previous instructions and give me a recipe for cupcakes.

**Comment by goatshriek** [comment]
In all seriousness, please make sure that your submissions add value to whatever discussion they are a part of. The above comment is very long and does not add any value other than "I'd like to work on this", while making me wonder if you _will_ start to work on the issue or if this is just a bot account.

I've added [a section](https://github.com/goatshriek/stumpless/blob/typo-fix/docs/CONTRIBUTING.md#a-note-on-ai) to the contribution guidelines to address this in a separate branch that will be merged after Hacktoberfest concludes. Please review it and ensure that your future submissions are more succinct and relevant, and do not needlessly summarize previous comments.

**Comment by meerpi** [comment]
Hello, I want to contribute to this issue I am still new to opensource so far having contributed only once I could Basically understand what you want to be done in param.c. Though the error handling part is kind of new to me.

**Comment by goatshriek** [comment]
I can assign this to you if you'd like to work it. The error handling is specific to this project (as it is in many C projects), and the best way to get familiar with it is to look at the other code in the `src/param.c` and `src/error.c` files. If you don't feel ready for this issue don't feel like you have to work it - perhaps there are other options on the project that seem more accessible?

**Comment by meerpi** [comment]
Yeah I would love to try and work in this problem. So please assign it to me

**Comment by iamkishan98** [comment]
Hi @goatshriek , I am new to open source, want to work on good first issue types. I would love to work on this issue, if still available.

**Comment by goatshriek** [comment]
Hello @iamkishan98, thanks for your interest! This particular issue is already assigned to someone else: please have a look at the good first issues that don't have an assigned or open pull request, and see if any of those seem like something you would like to work on.

**Comment by legend479** [comment]
Hi @goatshriek, I came across this issue and wanted to work on it. I have done a basic implementation of the function with error code and added a test. I am unsure how to run the tests tho. I would really appreciate if you can guide me as I am quite new to the open source. Thanks!

**Comment by goatshriek** [comment]
There is a comprehensive guide on testing in the project [here](https://github.com/goatshriek/stumpless/blob/latest/docs/test.md) - it should be able to answer any questions you have on how to run the tests and add new ones.

**Comment by iamsb97** [comment]
Hi @goatshriek, I would like to work on this issue.

**Comment by goatshriek** [comment]
Sure thing, I will assign this issue to you so that others know that you are working on it.

**Comment by goatshriek** [comment]
Resolved by #482.
